File: android-beans/src/test/java/rocks/frieler/android/beans/BeanConfigurationsBeansCollectorHandlesBeanDependenciesTest.java
package rocks.frieler.android.beans;

import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java8.util.Optional;

import static org.hamcrest.CoreMatchers.hasItem;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.junit.Assert.assertThat;

/**
 * Tests the {@link BeanConfigurationsBeansCollector} in more complex situations where {@link BeanConfiguration}s
 * depend on each other.
 */
public class BeanConfigurationsBeansCollectorHandlesBeanDependenciesTest {
    private final BeanRegistry beanRegistry = new BeanRegistry();
    private BeanConfigurationsBeansCollector beanConfigurationsBeansCollector = new BeanConfigurationsBeansCollector(beanRegistry);

    @Test(expected = BeanInstantiationException.class)
    public void testCollectBeansFailsOnCyclicDependency() {
        BeanConfiguration beanConfiguration1 = new BeanConfiguration1(new SingleBeanDependency<>(BeanConfiguration2.class));
        BeanConfiguration beanConfiguration2 = new BeanConfiguration2(new SingleBeanDependency<>(BeanConfiguration1.class));

        beanConfigurationsBeansCollector.collectBeans(Arrays.asList(beanConfiguration1, beanConfiguration2));
    }

    @Test
    public void testCollectBeansResolvesSingleBeanDependenciesBetweenBeanConfigurations() {
        BeanConfiguration beanConfiguration1 = new BeanConfiguration1();
        BeanConfiguration beanConfiguration2 = new BeanConfiguration2(new SingleBeanDependency<>(BeanConfiguration1.class));
        BeanConfiguration beanConfiguration3 = new BeanConfiguration3(new SingleBeanDependency<>(BeanConfiguration2.class));

        beanConfigurationsBeansCollector.collectBeans(Arrays.asList(beanConfiguration2, beanConfiguration3, beanConfiguration1));

        assertThat(beanRegistry.lookUpBean(BeanConfiguration1.class), is(notNullValue()));
        assertThat(beanRegistry.lookUpBean(BeanConfiguration2.class), is(notNullValue()));
        assertThat(beanRegistry.lookUpBean(BeanConfiguration3.class), is(notNullValue()));
    }

    @Test
    public void testCollectBeansResolvesTransitiveOptionalDependencies() {
        final BeanConfiguration1 beanConfiguration1 = new BeanConfiguration1(new OptionalSingleBeanDependency<>(BeanConfiguration2.class)) {
            @Override
            public void defineBeans(BeansCollector beansCollector) {
                //noinspection unchecked
                assertThat(((Optional<?>) getDependencies().get(0).get()).isPresent(), is(true));
                super.defineBeans(beansCollector);
            }
        };
        final BeanConfiguration2 beanConfiguration2 = new BeanConfiguration2(new OptionalSingleBeanDependency<>(BeanConfiguration3.class)) {
            @Override
            public void defineBeans(BeansCollector beansCollector) {
                //noinspection unchecked
                assertThat(((Optional<?>) getDependencies().get(0).get()).isPresent(), is(true));
                super.defineBeans(beansCollector);
            }
        };
        final BeanConfiguration3 beanConfiguration3 = new BeanConfiguration3();

        beanConfigurationsBeansCollector.collectBeans(Arrays.asList(beanConfiguration1, beanConfiguration2, beanConfiguration3));

        assertThat(beanRegistry.lookUpBean(BeanConfiguration1.class), is(notNullValue()));
        assertThat(beanRegistry.lookUpBean(BeanConfiguration2.class), is(notNullValue()));
        assertThat(beanRegistry.lookUpBean(BeanConfiguration3.class), is(notNullValue()));
    }

    @Test
    public void testCollectBeansResolvesDependenciesBetweenBeanConfigurationsWhenOneRequestsAllBeansOfAType() {
        BeanConfigurationNeedingAllBeansOfTypeBeanConfiguration1 beanConfigurationNeedingAllBeansOfTypeBeanConfiguration1 = new BeanConfigurationNeedingAllBeansOfTypeBeanConfiguration1();
        BeanConfiguration1 beanConfiguration1 = new BeanConfiguration1();
        BeanConfiguration beanConfiguration2 = new BeanConfiguration2();
        BeanConfiguration beanConfiguration3DependingOn2 = new BeanConfiguration3(new SingleBeanDependency<>(BeanConfiguration2.class));

        beanConfigurationsBeansCollector.collectBeans(Arrays.asList(
                beanConfigurationNeedingAllBeansOfTypeBeanConfiguration1,
                beanConfiguration3DependingOn2,
                beanConfiguration1,
                beanConfiguration2));

        assertThat(beanRegistry.lookUpBean(BeanConfiguration1.class), is(notNullValue()));
        assertThat(beanRegistry.lookUpBean(BeanConfiguration2.class), is(notNullValue()));
        assertThat(beanRegistry.lookUpBean(BeanConfiguration3.class), is(notNullValue()));
        assertThat(beanRegistry.lookUpBean(BeanConfigurationNeedingAllBeansOfTypeBeanConfiguration1.class), is(notNullValue()));
        assertThat(beanConfigurationNeedingAllBeansOfTypeBeanConfiguration1.beanConfiguration1s, hasItem(beanConfiguration1));
    }


    private static abstract class BeanConfigurationWithDefinedDependencies extends BeanConfiguration {
        private final List<BeanDependency<?>> beanDependencies;

        private BeanConfigurationWithDefinedDependencies(BeanDependency<?>... beanDependencies) {
            this.beanDependencies = Arrays.asList(beanDependencies);
        }

        @Override
        protected List<BeanDependency<?>> getDependencies() {
            return beanDependencies;
        }
    }

    private static class BeanConfiguration1 extends BeanConfigurationWithDefinedDependencies {
        private BeanConfiguration1(BeanDependency<?>... beanDependencies) {
            super(beanDependencies);
        }

        @Override
        public void defineBeans(BeansCollector beansCollector) {
            beansCollector.defineBean(this);
        }
    }

    private static class BeanConfiguration2 extends BeanConfigurationWithDefinedDependencies {
        private BeanConfiguration2(BeanDependency<?>... beanDependencies) {
            super(beanDependencies);
        }

        @Override
        public void defineBeans(BeansCollector beansCollector) {
            beansCollector.defineBean(this);
        }
    }

    private static class BeanConfiguration3 extends BeanConfigurationWithDefinedDependencies {
        private BeanConfiguration3(BeanDependency<?>... beanDependencies) {
            super(beanDependencies);
        }

        @Override
        public void defineBeans(BeansCollector beansCollector) {
            beansCollector.defineBean(this);
        }
    }

    private static class BeanConfigurationNeedingAllBeansOfTypeBeanConfiguration1 extends BeanConfiguration {
        private BeanDependency<List<BeanConfiguration1>> beanConfiguration1sDependency = requireBeans(BeanConfiguration1.class);
        private List<BeanConfiguration1> beanConfiguration1s;

        @Override
        public void defineBeans(BeansCollector beansCollector) {
            beanConfiguration1s = beanConfiguration1sDependency.get();
            beansCollector.defineBean(this);
        }
    }
}

File: android-beans/src/test/kotlin/rocks/frieler/android/beans/BeanConfigurationsBeansCollectorHandlesBeanDependenciesTest.kt
package rocks.frieler.android.beans

import assertk.Assert
import assertk.assertThat
import assertk.assertions.contains
import assertk.assertions.isNotNull
import assertk.assertions.support.expected
import java8.util.Optional
import org.junit.Test

/**
 * Tests the [BeanConfigurationsBeansCollector] in more complex situations where [BeanConfiguration]s
 * depend on each other.
 */
class BeanConfigurationsBeansCollectorHandlesBeanDependenciesTest {
	private val beanRegistry = BeanRegistry()
	private val beanConfigurationsBeansCollector = BeanConfigurationsBeansCollector(beanRegistry)

	@Test(expected = BeanInstantiationException::class)
	fun `collectBeans() fails on cyclic dependency`() {
		val beanConfiguration1 = BeanConfiguration1(SingleBeanDependency(BeanConfiguration2::class.java))
		val beanConfiguration2 = BeanConfiguration2(SingleBeanDependency(BeanConfiguration1::class.java))

		beanConfigurationsBeansCollector.collectBeans(listOf(beanConfiguration1, beanConfiguration2))
	}

	@Test
	fun `collectBeans() resolves SingleBeanDependencies between BeanConfigurations`() {
		val beanConfiguration1 = BeanConfiguration1()
		val beanConfiguration2 = BeanConfiguration2(SingleBeanDependency(BeanConfiguration1::class.java))
		val beanConfiguration3 = BeanConfiguration3(SingleBeanDependency(BeanConfiguration2::class.java))

		beanConfigurationsBeansCollector.collectBeans(listOf(beanConfiguration2, beanConfiguration3, beanConfiguration1))

		assertThat(beanRegistry.lookUpBean(BeanConfiguration1::class.java)).isNotNull()
		assertThat(beanRegistry.lookUpBean(BeanConfiguration2::class.java)).isNotNull()
		assertThat(beanRegistry.lookUpBean(BeanConfiguration3::class.java)).isNotNull()
	}

	@Test
	fun `collectBeans() resolves transitive optional dependencies`() {
		val beanConfiguration1 = object : BeanConfiguration1(OptionalSingleBeanDependency(BeanConfiguration2::class.java)) {
			override fun defineBeans(beansCollector: BeansCollector) {
				assertThat((dependencies[0].get() as Optional<*>)).isPresent()
				super.defineBeans(beansCollector)
			}
		}
		val beanConfiguration2 = object : BeanConfiguration2(OptionalSingleBeanDependency(BeanConfiguration3::class.java)) {
			override fun defineBeans(beansCollector: BeansCollector) {
				assertThat((dependencies[0].get() as Optional<*>)).isPresent()
				super.defineBeans(beansCollector)
			}
		}
		val beanConfiguration3 = BeanConfiguration3()

		beanConfigurationsBeansCollector.collectBeans(listOf(beanConfiguration1, beanConfiguration2, beanConfiguration3))

		assertThat(beanRegistry.lookUpBean(BeanConfiguration1::class.java)).isNotNull()
		assertThat(beanRegistry.lookUpBean(BeanConfiguration2::class.java)).isNotNull()
		assertThat(beanRegistry.lookUpBean(BeanConfiguration3::class.java)).isNotNull()
	}

	@Test
	fun `collectBeans() resolves dependencies between BeanConfigurations when one requests all beans of a type`() {
		val beanConfigurationNeedingAllBeansOfTypeBeanConfiguration1 = BeanConfigurationNeedingAllBeansOfTypeBeanConfiguration1()
		val beanConfiguration1 = BeanConfiguration1()
		val beanConfiguration2: BeanConfiguration = BeanConfiguration2()
		val beanConfiguration3DependingOn2: BeanConfiguration = BeanConfiguration3(SingleBeanDependency(BeanConfiguration2::class.java))

		beanConfigurationsBeansCollector.collectBeans(listOf(
				beanConfigurationNeedingAllBeansOfTypeBeanConfiguration1,
				beanConfiguration3DependingOn2,
				beanConfiguration1,
				beanConfiguration2))

		assertThat(beanRegistry.lookUpBean(BeanConfiguration1::class.java)).isNotNull()
		assertThat(beanRegistry.lookUpBean(BeanConfiguration2::class.java)).isNotNull()
		assertThat(beanRegistry.lookUpBean(BeanConfiguration3::class.java)).isNotNull()
		assertThat(beanRegistry.lookUpBean(BeanConfigurationNeedingAllBeansOfTypeBeanConfiguration1::class.java)).isNotNull()
		assertThat(beanConfigurationNeedingAllBeansOfTypeBeanConfiguration1.beanConfiguration1s).contains(beanConfiguration1)
	}

	private abstract class BeanConfigurationWithDefinedDependencies(vararg beanDependencies: BeanDependency<*>) : BeanConfiguration() {
		init {
			listOf(*beanDependencies).forEach { dependency -> addDependency(dependency) }
		}
	}

	private open class BeanConfiguration1(vararg beanDependencies: BeanDependency<*>) : BeanConfigurationWithDefinedDependencies(*beanDependencies) {
		override fun defineBeans(beansCollector: BeansCollector) {
			beansCollector.defineBean(this)
		}
	}

	private open class BeanConfiguration2(vararg beanDependencies: BeanDependency<*>) : BeanConfigurationWithDefinedDependencies(*beanDependencies) {
		override fun defineBeans(beansCollector: BeansCollector) {
			beansCollector.defineBean(this)
		}
	}

	private class BeanConfiguration3(vararg beanDependencies: BeanDependency<*>) : BeanConfigurationWithDefinedDependencies(*beanDependencies) {
		override fun defineBeans(beansCollector: BeansCollector) {
			beansCollector.defineBean(this)
		}
	}

	private class BeanConfigurationNeedingAllBeansOfTypeBeanConfiguration1 : BeanConfiguration() {
		private val beanConfiguration1sDependency: BeanDependency<List<BeanConfiguration1>> = requireBeans(BeanConfiguration1::class.java)

		lateinit var beanConfiguration1s: List<BeanConfiguration1>

		override fun defineBeans(beansCollector: BeansCollector) {
			beanConfiguration1s = beanConfiguration1sDependency.get()!!
			beansCollector.defineBean(this)
		}
	}
}

fun Assert<Optional<*>>.isPresent() = transform { actual ->
	if (!actual.isPresent) {
		expected("to be present")
	}
}
